REGRESSION (250414@main): Placeholder text is vertically off center on mail.163.com

pxlcoder opened the pull request:
#### 0f760ccf1c9e695d6ecf5d6a215a07adc08dcd7f
<pre>
REGRESSION (250414@main): Placeholder text is vertically off center on mail.163.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=243966">https://bugs.webkit.org/show_bug.cgi?id=243966</a>
rdar://96399025

Reviewed by Ryosuke Niwa.

mail.163.com sets a height and equivalent line-height on `::placeholder`.
Following 250414@main, the line-height property cannot be set by the author,
matching the behavior of other browsers. However, this change exposed a flaw in
the logic that vertically centers `::placeholder`.

As of 222311@main, `::placeholder` is centered using its box&apos;s logical height.
However, if the height is larger than the height of content in the first line
box, and the line height does not match the logical height, the placeholder
text will appear vertically off center, even though the box itself is center
aligned.

To fix, use the height of the content in the placeholder&apos;s first line box to
determine its position. This approach works as the placeholder in input elements
is restricted to a single line. Additionally, prevent authors from changing the
vertical padding on `::placeholder`. This behavior matches other browsers, and
ensures the line box height approach keeps the placeholder text centered.

* LayoutTests/fast/forms/placeholder-content-center.html:

Augmented an existing reference test to verify that the placeholder text remains
centered when height and vertical padding are specified.

* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::TextControlPlaceholderElement::resolveCustomStyle):
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.h:
(WebCore::InlineIterator::LineBox::contentLogicalHeight const):
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(WebCore::RenderTextControlSingleLine::layout):

Canonical link: <a href="https://commits.webkit.org/253500@main">https://commits.webkit.org/253500@main</a>
</pre>

rniwa commented on Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
```diff
@@ -198,7 +199,9 @@ void RenderTextControlSingleLine::layout()
         if (containerBox) {
             // Center vertical align the placeholder content.
-            auto logicalTop = placeholderTopLeft.y() + (containerBox->logicalHeight() / 2 - placeholderBox->logicalHeight() / 2);
+            auto placeholderLineBox = InlineIterator::firstLineBoxFor(downcast<RenderBlockFlow>(*placeholderBox));
```
Are you sure getting the height of the first line box is always enough?
What if placeholder had a mixture of English & Japanese/Chinese?
Do we still use the correct height?

pxlcoder replied:
I tested with a mixture of English/Chinese characters, and it appears to layout correctly.

Is there a specific behavior you’re concerned about?

(cc @alanbujtas)

rniwa replied:
Okay. Have you tried Hebrew / Arabic & English? I just wanna make sure that just using the first box doesn't result in misalignment.

pxlcoder replied:
Arabic/English looks fine too.

<img width="353" alt="Screenshot 2022-08-16 at 10 36 29 AM" src="https://user-images.githubusercontent.com/1904138/184943593-fc739249-0273-4ee1-98b4-0c3ab177910d.png">

rniwa replied:
Very nice!

alanbaradlay replied:
what kind of misalignment are you thinking of? The line box by definition contains all the inline content for that particular line regardless of their attributes etc. The only way this may fail if  the "placeholder" text wrapped into multiple lines and for that the content height would be: (lastLineBox.contentLogicalBottom() - firsLineBox.contentLogicalTop()). As for the actual height value, this patch specifically calls into the _content_logicalHeight() API which is the enclosing geometry for the line (and not the line box's height value that we use to size the containing block container). The commit message should be a bit more clear about that actually.
"use the placeholder's first line box height "
This is not the line box height, this is the height of the content inside the first line box.

rniwa replied:
Line boxes can be split into multiple boxes with bidirectional text.

alanbaradlay replied:
how do you split a line box with bidirectional text? are you referring to this line box https://www.w3.org/TR/css-inline-3/#line-box or something else (maybe inline box)?

alanbaradlay replied:
The "first box" refers to a _line_ box, not the content inside the line e.g. an inline box.

rniwa replied:
Oh I see. Sorry, I got confused between line boxes & inline boxes.

pxlcoder committed:
REGRESSION (250414@main): Placeholder text is vertically off center on mail.163.com
https://bugs.webkit.org/show_bug.cgi?id=243966
rdar://96399025

Reviewed by Ryosuke Niwa.

mail.163.com sets a height and equivalent line-height on `::placeholder`.
Following 250414@main, the line-height property cannot be set by the author,
matching the behavior of other browsers. However, this change exposed a flaw in
the logic that vertically centers `::placeholder`.

As of 222311@main, `::placeholder` is centered using its box's logical height.
However, if the height is larger than the height of content in the first line
box, and the line height does not match the logical height, the placeholder
text will appear vertically off center, even though the box itself is center
aligned.

To fix, use the height of the content in the placeholder's first line box to
determine its position. This approach works as the placeholder in input elements
is restricted to a single line. Additionally, prevent authors from changing the
vertical padding on `::placeholder`. This behavior matches other browsers, and
ensures the line box height approach keeps the placeholder text centered.

* LayoutTests/fast/forms/placeholder-content-center.html:

Augmented an existing reference test to verify that the placeholder text remains
centered when height and vertical padding are specified.

* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::TextControlPlaceholderElement::resolveCustomStyle):
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.h:
(WebCore::InlineIterator::LineBox::contentLogicalHeight const):
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(WebCore::RenderTextControlSingleLine::layout):

Canonical link: https://commits.webkit.org/253500@main
